WordPress/gutenberg
# InputControl: Reverting recent changes, which affected updates

This update reverts 2 recent updates on the `InputControl` component:

https://github.com/WordPress/gutenberg/pull/25609
https://github.com/WordPress/gutenberg/pull/25753

I've noticed some issues in the BoxControl component in Gutenberg with the updates. In particular...

* adding new values and pressing `Tab` to commit
* drag to update no longer works :(

<img width="306" alt="Screen Shot 2020-10-07 at 12 08 36 PM" src="https://user-images.githubusercontent.com/2322354/95357435-d9abdb00-0895-11eb-8fd5-0e76950fd771.png">

cc'ing @stokesman 

Looking back at the initial issue (https://github.com/WordPress/gutenberg/issues/24460), these updates are attempting to fix some number input entering.

~I haven't been able to replicate the issues. Maybe I'm not entering things correctly (to break it) in my keyboard?~
~I'd love to know more info about this.~
~On top of reverts, I'd like to add solutions for the issues mentioned in https://github.com/WordPress/gutenberg/issues/24460~

## Updates

Beyond the revert, the changes can be seen here:

https://github.com/WordPress/gutenberg/pull/25913/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R66-R83

https://github.com/WordPress/gutenberg/pull/25913/files#diff-559f6ca9af4f175b7f7d4b1741a7f4d6R30-R42

### 📹 Demo + Walkthrough

I recorded a video demo + walkthrough that showcases the issue and describes the fixes:
https://www.loom.com/share/22b6d3b1b61140e4a667ed61a6858f52

### ⚛️ Fixing State Handling

The issue was that the inner number input wasn't correctly handling and coordinating values/changes between the UI element (`input[type="number"]`) and the state that exists within the wrapping `RangeControl` component.

The solution involved swapping `useControlledState` for a simple `useState` + `useEffect` hook combination for state management and syncing within the inner number `InputField`.

The solution also involved (re)adding the HTML validation step before "commiting" (firing the `onChange`) callback. This allows for "invalid" state to be held within the number input before pushing it upwards.

What this means, is that users will be able to **type** in values like `123`, when the `min` attribute is set to `20`.

Previously, it would **jump** to `20` because typing in the initial character (`1`) would trigger onChange immediately... which clamps the value and propagates it updates.

This is no longer the case! (Thank goodness)

### 🗜 Clamping onChange

Secondly, the final value that is sent from the `RangeControl` onChange callback is clamped. Previously, this was not the case. We also didn't spot this problem due to the above mentioned issue.

## How has this been tested?

* Confirm reverting the commits fixes BoxControl issues and overall drag to update interaction
* Confirm `RangeControl` state flow is working

(Note: For drag to update, I think order matters for the `onMouseDown` callback. This isn't as tricky as the update that adjusts the update mechanism to rely on the `blur` event)


## How to test

* Run `npm run dev`
* Add a Cover block
* Adjust the height using the input control. This should work.
* Add a Columns block
* Interfacing with the RangeControl + accompanying number input should work.

## Checklist:
- [n/a] My code follows the WordPress code style. 
- [n/a] My code follows the accessibility standards. 
- [n/a] My code has proper inline documentation. 
- [n/a] I've included developer documentation if appropriate. 
- [n/a] I've updated all React Native files affected by any refactorings/renamings in this PR.
